sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -477,6 +478,25 @@ object RemoveRedundantAliases extends Rule[LogicalPlan] {
   def apply(plan: LogicalPlan): LogicalPlan = removeRedundantAliases(plan, AttributeSet.empty)
 }
 
+/**
+ * Remove redundant aggregates from a query plan. A redundant aggregate is an aggregate whose
+ * only goal is to keep distinct values, while its parent aggregate would ignore duplicate values.
+ */
+object RemoveRedundantAggregates extends Rule[LogicalPlan] {
+  def apply(plan: LogicalPlan): LogicalPlan = plan transformUp {
```

Comment:
`transformUp` -> `transform`? Seems like this transformation does not depend on the order.

Reply:
There is a slight difference - `transformUp` can handle more than two consecutive aggregates in a single iteration.
If there are 3 aggregates `A1(A2(A3(...)))`, where `A2` and  `A3` are redundant, then with `transformDown` on first iteration `A2` is removed and then `A3` on the second iteration. With `transformUp` it removes `A3` and then `A2` on the same iteration.

Reply:
>> where A2 and A3 are redundant

If only A2 and A3 are redundant, `lowerIsRedundant(A1, A2)` returns false and it seems like A2/A3 are removed in the same iteration even in `transform` (the first acse). I miss something?

Reply:
I meant that `lowerIsRedundant(A1, A2)` would return true,

Reply:
> transformUp can handle more than two consecutive aggregates in a single iteration.

You meant that `transformUp` can transform `A1(A2(A3(...))) => A1()` in a single iteration? Probably, I miss some conditions for the case you described above, so could you give me a concrete query example for it?

Reply:
For example the UT `Remove 2 redundant aggregates`

Reply:
Ah, ok. I got your point. Looks okay as it is.